relayer/canonical_validator_client.go

```diff
@@ -34,10 +34,51 @@ func (v *CanonicalValidatorClient) GetSubnetID(ctx context.Context, chainID ids.
 	return v.client.ValidatedBy(ctx, chainID)
 }
 
+// Gets the current validator set of the given subnet ID, include the validators' BLS public keys.
+// This implementation of GetValidatorSet currently makes two RPC requests, one to get the
+// subnet validators, and another to get their BLS public keys. This is necessary in order to enable
+// the use of the public APIs (which don't support "GetValidatorsAt") because BLS keys are currently
+// only associated with primary network validation periods. If ACP-13 is implementated in the future
+// (https://github.com/avalanche-foundation/ACPs/blob/main/ACPs/13-subnet-only-validators.md), it may
+// become possible to reduce this to a single RPC request that returns both the subnet validators
+// as well as their BLS public keys.
 func (v *CanonicalValidatorClient) GetValidatorSet(
 	ctx context.Context,
 	height uint64,
 	subnetID ids.ID,
 ) (map[ids.NodeID]*validators.GetValidatorOutput, error) {
-	return v.client.GetValidatorsAt(ctx, subnetID, height)
+	// Get the current subnet validators. These validators are not expected to include
+	// BLS signing information given that addPermissionlessValidatorTx is only used to
+	// add primary network validators.
+	subnetVdrs, err := v.client.GetCurrentValidators(ctx, subnetID, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Look up the primary network validators of the NodeIDs validating the subnet
+	// in order to get their BLS keys.
+	res := make(map[ids.NodeID]*validators.GetValidatorOutput, len(subnetVdrs))
+	subnetNodeIDs := make([]ids.NodeID, 0, len(subnetVdrs))
+	for _, subnetVdr := range subnetVdrs {
+		subnetNodeIDs = append(subnetNodeIDs, subnetVdr.NodeID)
+		res[subnetVdr.NodeID] = &validators.GetValidatorOutput{
+			NodeID: subnetVdr.NodeID,
+			Weight: subnetVdr.Weight,
+		}
+	}
+	primaryVdrs, err := v.client.GetCurrentValidators(ctx, ids.Empty, subnetNodeIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set the BLS keys of the result.
+	for _, primaryVdr := range primaryVdrs {
+		vdr, ok := res[primaryVdr.NodeID]
```

Comment:
Are we ever expecting `primaryVdr` to not be found in `res`? My understanding is that the above call to `GetCurrentValidators` applies this filter already.

Reply:
Nope, I don't think we ever expect it currently. I'll add a warning log for this case. I think it still makes sense to `continue` rather than returning an error, because the relayer should be able to handle this case the same as if the given validator didn't register a BLS public key on the P-chain. As long as a sufficient % of the subnet stake weight has BLS keys registered and set here, the message should be able to be successfully relayed still.

Reply:
Yup, I agree.

Can we add a check below that `primaryVdr.Signer != nil`? And if it is, we can emit a warning and `continue` like we do here.

Reply:
Yep, I think that proper check is already there now actually. I chose not to emit any log in that case because it's expected (or at least certainly possible) for the primary network validator to not have any BLS public key registered.

Reply:
Let me know if you feel strongly otherwise though. I'm neither here or there on the log 👍

Reply:
I may be missing something, but it looks like the first time we access primaryVdr.Signer it's unchecked. If it's `nil`, that would cause the app to panic.